api/docs/guides/data-migrations.md

```diff
@@ -0,0 +1,90 @@
+1. Use a data migration management command to move data from the previous column to the new column. To find the rows that need updating, iterate through the table by querying for rows that do not have a value in the new column yet. Because the version of the application running at this point is writing and reading from the new column (falling back to the old for reads when necessary), the query will eventually return zero rows.
+1. Once the data migration is complete, deploy a new version of the application that removes the old column and the fallback reads to it and only uses the new column.
+
+To reiterate, yes, this is a much more tedious process. However, note the following benefits.
+
+### Benefits of this approach
+
+#### Zero-downtime
+
+The entire point, of course. This benefits everyone who depends on the application's uptime and reliability.
+
+#### Reversibility
+
+If the new version of the application has a critical bug, whether related to the data changes or not, we can revert each step to the previous version without issue or data loss. Even during the data migration process, because the version of the application running is updating both columns, if you have to revert to the first version (or even earlier) that doesn't use the new column, the old column will still have up-to-date data and no user data will be lost. This would complicate the data migration process, however, as previous versions of the application will not be updating the new column and would likely require deleting the data from the new column to start the data migration process over from the start. This can cause massive time consumption but is overall less of a headache than data loss or fully broken deployments.
+
+#### Intentionality and expediency
+
+Due to the great lengths it takes to change a column name, the process will inevitably cause contributors to ask themselves: is this worth it? While changing the name of a column can be helpful to disambiguate the data in the column, using a model attribute alias can be just as helpful without any of the disruption or time of a data migration. These kinds of questions prompt us to make expedient choices that deliver features, bug fixes, and developer experience improvements faster.
+
+## Django management command data migrations
+
+### Why use management commands for data migrations instead of SQL?
+
+Django comes with a data migration feature built in that allows executing data transformations using SQL. If you want to move data between two columns, it is trivial to do so with SQL and Django makes it just as easy. [Documentation for this Django feature is available here](https://docs.djangoproject.com/en/4.1/topics/migrations/#data-migrations).
+
+When considering the potential issues with using SQL data migrations with our current deployment strategy, keep in mind the following details:
+
```

Comment:
I have a doubt about this section. The title and subtitle make it sound like Django management commands are the solution to the problems described here, but I'm not sure how that is the case. After all, DB-related commands are mostly plain SQL under the hood.
I can see the Django commands as an easier way to handle migrations automatically for our future improved CI/CD pipeline but how are they supposed to solve the problems of time?

Reply:
I haven't finished writing this section yet. The reason why management commands solve the issue at hand (long-running data migrations), isn't because they make data migrations _shorter_ or not use SQL. Of course, as you said, they still use SQL as they'll still be manipulating the database via the same tools that a migration file would. The difference is that a management command allows us to break up the SQL commands to iterate over the set of data that needs to be transformed and critically _not at application startup_. Additionally, they have the following features:

1. They can be stopped and started to allow for other more critical operations to take precedence or to recover from unexpected errors
2. They can be metered to prevent creating significant read/write lag that would hurt the overall application performance

Both of these _might_ be doable in pure SQL (maybe through a stored procedure, for example), but it's more effort, harder to test and have standardised tools, and I don't think we have great tools at the moment for creating or maintaining stored procedures in the API anyway.

Management commands solve these problems in a familiar setting, with testing tools that we already know how to use and for which we have invested time into building utilities like data fixtures and helpers.

The main benefit, though, is being able to break up the transformation into smaller, iterable chunks, that can be executed outside of application startup time (which we can't do with Django migrations without making significant complications to our deployment workflow).